device/Microsoft.Azure.Devices.Client/Common/TaskHelpers.cs

```diff
@@ -384,20 +390,25 @@ public static Task WithTimeout(this Task task, TimeSpan timeout, Func<string> er
 
             if (task.IsCompleted || (timeout == Timeout.InfiniteTimeSpan && token == CancellationToken.None))
             {
-                await task;
+                await task.ConfigureAwait(false);
                 return;
             }
 
-            using (var cts = CancellationTokenSource.CreateLinkedTokenSource(token))
+            using (var delayCts = new CancellationTokenSource())
```

Comment:
>new CancellationTokenSource() [](start = 34, length = 29)

why do you remove CreateLinkedTokenSource method call from line 397? if input cancellation token in line 384 is canceled from caller, line 399 still need to wait till timeout trigger.

Reply:
That was the bug: there are 2 separate tokens here:
- delayCts is used to cancel the Task.Delay
- operationCts is used to attach to the token given by the user

[Here](https://github.com/Azure/azure-iot-sdk-csharp/pull/285/files#diff-5c60a411b2230a22a19ebff0240453beR401) the user operation completed before the Task.Delay and so we are cancelling the delay and returning the task result (by awaiting `task`).

[Here](https://github.com/Azure/azure-iot-sdk-csharp/pull/285/files#diff-5c60a411b2230a22a19ebff0240453beR407) the delay completed before the user operation and so we are cancelling the user operation (based on the `token` passed as argument [here](https://github.com/Azure/azure-iot-sdk-csharp/pull/285/files#diff-5c60a411b2230a22a19ebff0240453beL378)) then throwing the TimeoutException.

The original code was incorrectly combining the two processes into a single call which meant that if the operation completed successfully, the code would then try to cancel it [here](https://github.com/Azure/azure-iot-sdk-csharp/pull/285/files#diff-5c60a411b2230a22a19ebff0240453beL395).

Reply:
The passed in `token` in these helpers represents the same token passed to the original operation and so when it's cancelled, the operation is cancelled. `task` will complete in this case which means `delayCts` will cancel the Delay task.

The whole `TaskHelpers` is a mess and must be removed in the future completely. The case you're describing is the way a task should be cancelled: only through the cancellationToken and not by starting another delay task. CancellationTokenSource supports the timeout mechanism.

Reply:
@jasmineymlo I misunderstood CreateLinkedTokenSource(token). I originally thought calling the linked cts will cancel `token` but instead the linked cts will be set to cancelled.
The leak is still present with infinite retries because of this (the operation is not actually cancelled...). I need to figure out a way to cancel the original token when the operation times out.